### PR TITLE
Water ssr

### PR DESCRIPTION
**Pull request by e2002e**
Cleaned up.

### CODE REVIEW

**Comment by e2002e** [comment]
This has been a bit long, you should get that it's a bit like I am holding my breath seeing if change gets merged.

**Comment by e2002e** [comment]
Did you test that ? There's an option for Ssr in world's ocean and it works great @MoritzBrueckner.

**Comment by MoritzBrueckner** [comment]
Hi, I usually let Lubos decide whether he want's to merge something a little more substantial than typos – I may have the necessary permissions for this repository but he is the owner of this project.

I now tried your changes and I get this error:

```
ERROR: [...]\compiled\Shaders\water_pass.frag.glsl:72: 'ssrSearchDist' : undeclared identifier
ERROR: [...]\compiled\Shaders\water_pass.frag.glsl:72: '' : missing #endif
ERROR: [...]\compiled\Shaders\water_pass.frag.glsl:72: '' : compilation terminated
ERROR: 3 compilation errors.  No code generated.
```

It doesn't matter whether the water's SSR option is on or off, the error happens in both cases. It goes away when I enable SSR in the renderpath settings.

Also, if both SSR options are activated I still don't see any reflections:

Screenshot

Can you change "Ssr" in the UI to "SSR"?

**Comment by e2002e** [comment]
OK, thanks for testing.

**Comment by e2002e** [comment]
Set now. I discared the options ssr for water, it will be activated alonside the option in postprocess.
Screenshot_20220829_142501

**Comment by MoritzBrueckner** [comment]
Can confirm that it works now :) However, the water's normals are currently ignored by the SSR (it instead reads the normal from the gbuffer which you don't need since the water has known constant normals all the time, it's just a flat plane), so maybe you can change this.

Also I guess you don't need the changes in `props_renderpath.py` and `write_data.py` anymore since your latest changes.

**Comment by e2002e** [comment]
I've done as required, reflection though, doesn't apply when no object is below. Understandable for refraction but there I can't tell.
Capture d’écran du 2022-09-02 19-54-26

**Comment by MoritzBrueckner** [comment]
Nice 👍 There is a semicolon missing in water_pass.frag.glsl in line 150, but after adding it it works as expected :)

File: Sources/armory/renderpath/RenderPathDeferred.hx [hunk 1]
```diff
@@ -650,6 +649,7 @@ class RenderPathDeferred {
 			path.drawShader("shader_datas/copy_pass/copy_pass");
 			path.setTarget("tex");
 			path.bindTarget("_main", "gbufferD");
+			path.bindTarget("gbuffer0", "gbuffer0");
```
Comment (MoritzBrueckner) [inline Sources/armory/renderpath/RenderPathDeferred.hx:652]: Is this line still required?

File: Shaders/water_pass/water_pass.json [hunk 2]
```diff
@@ -56,6 +56,14 @@
...
 					"link": "_envmapRadiance",
 					"ifdef": ["_Rad"]
+				},
+				{
+					"name": "P",
+					"link": "_projectionMatrix"
+				},
+				{
+					"name": "V3",
+					"link": "_viewMatrix3"
```
Comment (luboslenco) [inline Shaders/water_pass/water_pass.json:66]: Looks like you can condition these with `"ifdef": ["_SSR"]` here? 🙂

**Comment by e2002e** [comment]
I took note of the requested changes, good to go now.
Thanks.

File: Shaders/water_pass/water_pass.frag.glsl [hunk 3]
```diff
@@ -9,10 +9,17 @@ uniform sampler2D tex;
 uniform sampler2D sbase;
 uniform sampler2D sdetail;
 uniform sampler2D sfoam;
+uniform mat4 P;
+uniform mat3 V3;
```
Comment (luboslenco) [inline Shaders/water_pass/water_pass.frag.glsl:13]: Last one - please add `#ifdef _SSR` here, not necessary but helps to tell it's for the SSR feature.

File: Shaders/water_pass/water_pass.frag.glsl [hunk 4]
```diff
@@ -9,10 +9,17 @@ uniform sampler2D tex;
...
 #ifdef _Rad
 uniform sampler2D senvmapRadiance;
 #endif
 
+#ifdef _CPostprocess
+uniform vec3 PPComp9;
+uniform vec3 PPComp10;
+#endif
```
Comment (luboslenco) [inline Shaders/water_pass/water_pass.frag.glsl:21]: Same here.

**Comment by e2002e** [comment]
Ach ich habe das vergesehen, es ist alles gut ich glaube.

**Comment by luboslenco** [comment]
Danke! Fixed one more typo at https://github.com/armory3d/armory/commit/5ff91eeddc7eb570a33565ee2a267c9948260086.

**Pull request merged**